_posts/2021-03-19-forecasting-time-series-with-torch_4/forecasting_4.Rmd

```diff
@@ -0,0 +1,570 @@
+
+Today, we don't replace *seq2seq* by something completely different. Instead, we enrich it, adding in a new component: the so-called *attention* module. Originally introduced around 2014[^1], attention mechanisms have gained enormous traction, so much so that a recent paper title starts out "Attention is Not All You Need"[^2].
+
+[^1]: For important papers see: @BahdanauCB14, @VinyalsKKPSH14, @XuBKCCSZB15.
+
+[^2]: @2021arXiv210303404D, thus replying to @2017arXiv170603762V, whose 2017 paper "went viral" for stating the opposite.
+
+The idea is the following. In the classic encoder-decoder setup, the decoder gets "primed" with an encoder summary just a single time: the time it starts its forecasting loop. From then on, it's on its own. With attention, however, it gets to see the complete sequence of encoder outputs again every time it forecasts a new value. And more than that: Every time, it gets to zoom in on *those* outputs that seem *relevant for the current prediction step*.
+
+We can easily see how this is a useful strategy in translation: In generating the next word, a model will need to know what part of the source sentence to focus on. How much the technique helps with numerical sequences, in contrast, will likely depend on the features of the series in question. Meaning: Even if, in our current case, it should turn out to have little effect, it will still be worthwhile just to have learned about the technique.
+
+# Data input
+
+In this post, as before, we work with `vic_elec`, but this time, we partly deviate from the way we used to employ it. With the original, bi-hourly dataset, training the current model takes a long time, longer than readers will want to wait when experimenting. So instead, we aggregate observations by day. In order not too have too few data, we train on years 2012 and 2013, reserving 2014 for validation as well as post-training inspection.
+
+```{r}
+vic_elec_daily <- vic_elec %>%
+  select(Time, Demand) %>%
+  index_by(Date = date(Time)) %>%
+  summarise(
+    Demand = sum(Demand) / 1e3) 
+
+elec_train <- vic_elec_daily %>% 
+  filter(year(Date) %in% c(2012, 2013)) %>%
+  as_tibble() %>%
+  select(Demand) %>%
+  as.matrix()
+
+elec_valid <- vic_elec_daily %>% 
+  filter(year(Date) == 2014) %>%
+  as_tibble() %>%
+  select(Demand) %>%
+  as.matrix()
+
+elec_test <- vic_elec_daily %>% 
+  filter(year(Date) %in% c(2014), month(Date) %in% 1:4) %>%
+  as_tibble() %>%
+  select(Demand) %>%
+  as.matrix()
+
+train_mean <- mean(elec_train)
+train_sd <- sd(elec_train)
+
+```
+
+We'll attempt to forecast demand up to fourteen days ahead. How long, then, should be the input sequences? As usual, this is a matter of experimentation; all the more now that we're adding in the attention mechanism. (Intuitively, it might not handle very long sequences so well). Below, we go with fourteen days here too, but that may not necessarily be the best possible choice for this series.
```

Comment:
```suggestion
We'll attempt to forecast demand up to fourteen days ahead. How long, then, should be the input sequences? This is a matter of experimentation; all the more so now that we're adding in the attention mechanism. (Intuitively, it might not handle very long sequences so well). Below, we go with fourteen days here too, but that may not necessarily be the best possible choice for this series.
```

Reply:
> Below, we go with fourteen days here too, but that may not necessarily be the best possible choice for this series.

Is this for the input sequence? If so, clarify/make that explicit. It's a bit confusing since _here_ could refer to any of what you do in this post. 

_Intuitively_ is a bit like "easily" or "simply", etc.—if it's not self-evident to the reader (e.g. for me), it doesn't do anything other than make me think that I should know something that I don't. That parenthetical also seems a bit off in terms of its location.

Reply:
> Intuitively is a bit like "easily" or "simply", etc.—if it's not self-evident to the reader (e.g. for me),

Ooooh... good you're saying that ... this was not at all the intention

It really should convey "I suspect that ..." 
Will rephrase